pkg/agent/controller/service_import.go

```diff
@@ -442,3 +564,48 @@ func (c *ServiceImportController) localServiceImportLister(transform func(si *mc
 
 	return retList
 }
+
+func findClusterWithOldestTimestamp(from map[string]string) string {
```

Comment:
Am wondering if we can improve this logic. In a scale setup there might be too many clusters exporting a service and may cause issues loopin through all of them.

How about we also add an annotation like `timestamp.submariner.io/oldest` and use that to check if we're older or not. When the oldest one is exported, only then we will need to cycle through all to find the new oldest.

Also, this may have some issues when we're usin VIP. In case two different clusters export service at same time, VIP will end up being allocated by whosoever was first to successfully create Aggregated SI. It may happen that its timestamp is newer than the one that failed to create, coz it depends on timing of when the create attempt is made on broker.

Reply:
> Am wondering if we can improve this logic. In a scale setup there might be too many clusters exporting a service and may cause issues loopin through all of them.

I think the time would be negligible with a single loop even with thousands. If it was a nested loop, then we could have an issue. With a single loop, it would take millions of iterations to start to become significant.

> 
> How about we also add an annotation like `timestamp.submariner.io/oldest` and use that to check if we're older or not. When the oldest one is exported, only then we will need to cycle through all to find the new oldest.

yeah we could do that, ie also if the oldest cluster is un-exported then remove the annotation to trigger other clusters to re-evaluate. But it adds more complexity that I'm not sure we really need. I can try it out....

> 
> Also, this may have some issues when we're usin VIP. In case two different clusters export service at same time, VIP will end up being allocated by whosoever was first to successfully create Aggregated SI. It may happen that its timestamp is newer than the one that failed to create, coz it depends on timing of when the create attempt is made on broker.

I wasn't planning on using the oldest timestamp for the VIP b/c we won't ever re-allocate it once set. The first cluster to successfully create the aggregated SI sets the VIP.

Reply:
I did some testing of `findClusterWithOldestTimestamp` with varying number of entries in the annotations map:

- 1,000 entries took 38.793 µs
- 10,000 took 244.597 µs
- 1,000,000 took 46.399 ms
- 10,000,000 took 711.979 ms

I also tried calling `findClusterWithOldestTimestamp` in a nested loop. So calling it 1,000 times with 1,000 map entries took 21.924 ms (simulates 1000 services on 1000 clusters). With 10,000 and 10,000 it took 2.16 s, which is still pretty negligible.

So I think we should be fine the way it is.

Reply:
Would it be worth adding those benchmarks as a test?